### PR TITLE
Extra defensive programming for pub modification time assert

### PR DESCRIPTION
**Pull request by Hixie**
Let's see if this helps us track down this issue...

### CODE REVIEW

**Comment by Hixie** [comment]
I'll write a test and update this PR once https://github.com/flutter/flutter/pull/42369 lands.

File: packages/flutter_tools/lib/src/dart/pub.dart [hunk 1]
```diff
@@ -189,11 +191,32 @@ class _DefaultPub implements Pub {
...
+      throwToolExit('$directory: unexpected concurrent modification of pubspec.yaml while running pub.');
+    }
+    // We don't check if dotPackages was actually modified, because as far as we can tell sometimes
+    // pub will decide it does not need to actually modify it.
+    // Since we rely on the file having a more recent timestamp, though, we do manually force the
+    // file to be more recently modified.
+    final DateTime now = DateTime.now();
+    if (now.isBefore(originalPubspecYamlModificationTime)) {
+      printError(
+        'Warning: File "${fs.path.absolute(pubSpecYaml.path)}" was created in the future. '
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/dart/pub.dart]: Might want to include the unexpected future date in the error to give an extra hint.
Spooky! 👻

**Review by jmagman — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/dart/pub.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter_tools/lib/src/dart/pub.dart]: oh good idea

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by Hixie** [comment]
I made the change suggested by @jmagman and also added a test for that branch, which I'd missed before.

Thanks for the reviews, will land on green.

**Comment by jonahwilliams** [comment]
The last several failures have been this issue on the devicelab host. Landing to get the tree back to green

**Comment by jonahwilliams** [comment]
Looks like this got out of sync somehow

**Pull request merged**